common/src/primitives/amount.rs

```diff
@@ -29,11 +29,85 @@ pub struct Amount {
     val: IntType,
 }
 
+fn remove_right_most_zeros_and_decimal_point(s: String) -> String {
+    let point_pos = s.chars().position(|c| c == '.');
+    if point_pos.is_none() {
+        return s;
+    }
+    let s = s.trim_end_matches('0');
+    let s = s.trim_end_matches('.');
+    s.to_owned()
+}
+
 impl Amount {
     #[allow(dead_code)]
     pub fn new(v: u128) -> Self {
         Amount { val: v }
     }
+
+    pub fn into_fixedpoint_str(self, decimals: u8) -> String {
+        let amount_str = self.val.to_string();
+        let decimals = decimals as usize;
+        if amount_str.len() <= decimals {
+            let zeros =
+                "0".repeat(decimals - amount_str.len());
+            let result = "0.".to_owned() + &zeros + &amount_str;
+            
+            remove_right_most_zeros_and_decimal_point(result)
+        } else {
+            // insert decimal point at position number N from the right
+            let insert_pos = amount_str.len() - decimals;
+            let result = if insert_pos < amount_str.len() {
+                amount_str[..insert_pos].to_owned() + "." + &amount_str[insert_pos..]
+            } else {
+                amount_str
+            };
+            
+            remove_right_most_zeros_and_decimal_point(result)
+        }
```

Comment:
I deliberately avoided all numeric treatments before constructing the full string to simplify overflow and math error possibilities 😅. I won't defend my code much if you guys insist.

Reply:
I guess the main thing to watch out for is having too many decimals, so that `10^decimals` does not exceed the capacity of the underlying type. This check has to be done either way. I don't think there's anything else, have I missed something?